include/GafferScene/Constraint.h

```diff
@@ -98,7 +101,13 @@ class GAFFERSCENE_API Constraint : public SceneElementProcessor
 
 	private :
 
-		boost::optional<ScenePath> targetPath() const;
+		struct Target
+		{
+			ScenePath path;
+			const ScenePlug *scene;
+		};
+
+		boost::optional<Target> target() const;
```

Comment:
This seems fine.  I wonder if there would be a slightly cleaner version where we remove the external visibility of this stuff, and move target to an anonymous namespace in the cpp.  This would require the call site for target to look something like:

`target( targetPlug()->getValue(), targetScenePlug()->getInput() ? targetScenePlug() : inPlug(), ignoreMissingTargetPlug()->getValue() )`

... maybe that's a bit messy?  It's definitely not strictly better, but I think it was me, I would err on the side of keeping these details out of the header - I also just generally where possible like functions where their access to class scope variables is restricted to explicit arguments.  Not saying it's a good idea in this case, but it was the only thing that occured to me while reading this code.

Reply:
Yeah, I spent a bit of time wondering about this too. I didn't like the repetition of `targetPlug()->getValue(), targetScenePlug()->getInput() ? targetScenePlug() : inPlug()` though : I definitely wanted to make `target()` responsible for that logic. In the past I've tended to agree that explicit arguments are preferable to implicit access to the class, because it makes it easier to verify the implementation of `affects()`. But for more complex helpers I've found the explicit arguments too cumbersome, especially if you don't want to evaluate them all upfront because some are expensive but only necessary sometimes.

I think my preferred solution at the moment is what I did for the [`BranchCreator::parentPathsForSet()`](https://github.com/GafferHQ/gaffer/blob/master/include/GafferScene/BranchCreator.h#L204-L205) helper. There I added an accompanying `affectsParentPathsForSet()` query that was used in `affects()`. That meant that `parentPathsForSet()` could have a simpler, easier to call implementation, but it the logic for what it depended on was encapsulated next to it, rather than far away in `affects()`.

I wanted to take this latter approach here too, but it didn't quite fit the pattern, because `target()` returns a whole `ScenePlug` but only specific child plugs from that are actually used in hash/compute. Rather than agonise about that, I chose to leave things as they were. I think it's good enough for now, but I'm happy to continue the discussion with the aim of finding a simple pattern we can apply as widely as possible.